parser-typechecker/src/Unison/Codebase/Branch.hs

```diff
@@ -501,8 +510,16 @@ empty0 =
   Branch0 mempty mempty mempty mempty mempty mempty mempty mempty mempty mempty
 
 -- | Checks whether a Branch0 is empty.
+-- This is not as easy as checking whether the branch is equal to the `empty0` branch
+-- because child branches may be empty, but still have history.
 isEmpty0 :: Branch0 m -> Bool
-isEmpty0 = (== empty0)
+isEmpty0 (Branch0 _terms _types _children _edits deepTerms deepTypes deepTermMetadata deepTypeMetadata deepPaths deepEdits) =
+  Relation.null deepTerms
+    && Relation.null deepTypes
+    && Relation4.null deepTermMetadata
+    && Relation4.null deepTypeMetadata
+    && Set.null deepPaths
```

Comment:
what's `deepPaths` and should that be in here?

Reply:
It's just a list of all paths to child branches. It's handled 'correctly' with respect to deleted branches due to the updates to `deriveDeepPaths` I made in this PR, but it likely isn't adding much value being here.

Reply:
I'd probably remove it from this check then, as I think it's just generating confusion. :)

Reply:
I think `Relation4.null deepTermMetadata` and `Relation4.deepTypeMetadata` are implied by `Relation.null deepTypes` and `Relation.null deepTerms`.

Reply:
I agree, I left them in there just in case that changes in the future, but can remove them if that's preferred.
The performance change should be negligible anyways since `null` is fast, and the &&'s will short circuit most of the time.